File: lib/models/ball.dart
class Ball {
  final String id;
  final double x;
  final double y;
  final double radius;
  final double distanceToPiglet;
  final bool isPiglet;

  Ball({
    required this.id,
    required this.x,
    required this.y,
    required this.radius,
    this.distanceToPiglet = 0.0,
    this.isPiglet = false,
  });

  Ball copyWith({
    String? id,
    double? x,
    double? y,
    double? radius,
    double? distanceToPiglet,
    bool? isPiglet,
  }) {
    return Ball(
      id: id ?? this.id,
      x: x ?? this.x,
      y: y ?? this.y,
      radius: radius ?? this.radius,
      distanceToPiglet: distanceToPiglet ?? this.distanceToPiglet,
      isPiglet: isPiglet ?? this.isPiglet,
    );
  }

  Map<String, dynamic> toMap() => {
    'id': id,
    'x': x,
    'y': y,
    'radius': radius,
    'distanceToPiglet': distanceToPiglet,
    'isPiglet': isPiglet,
  };

  factory Ball.fromMap(Map<String, dynamic> map) => Ball(
    id: map['id'],
    x: map['x'],
    y: map['y'],
    radius: map['radius'],
    distanceToPiglet: map['distanceToPiglet'] ?? 0.0,
    isPiglet: map['isPiglet'] ?? false,
  );
}

File: py_boulometre/src/models/ball.py
"""
Modèle Ball pour PyBoul'O'Mètre
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class Ball:
    """Représente une boule ou un cochonnet"""
    id: str
    x: float
    y: float
    radius: float
    distance_to_piglet: float = 0.0
    is_piglet: bool = False
    
    def copy_with(
        self,
        id: Optional[str] = None,
        x: Optional[float] = None,
        y: Optional[float] = None,
        radius: Optional[float] = None,
        distance_to_piglet: Optional[float] = None,
        is_piglet: Optional[bool] = None
    ) -> 'Ball':
        """Crée une copie avec des valeurs modifiées"""
        return Ball(
            id=id if id is not None else self.id,
            x=x if x is not None else self.x,
            y=y if y is not None else self.y,
            radius=radius if radius is not None else self.radius,
            distance_to_piglet=distance_to_piglet if distance_to_piglet is not None else self.distance_to_piglet,
            is_piglet=is_piglet if is_piglet is not None else self.is_piglet
        )
    
    def to_dict(self) -> dict:
        """Convertit en dictionnaire"""
        return {
            'id': self.id,
            'x': self.x,
            'y': self.y,
            'radius': self.radius,
            'distance_to_piglet': self.distance_to_piglet,
            'is_piglet': self.is_piglet
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Ball':
        """Crée une Ball à partir d'un dictionnaire"""
        return cls(
            id=data.get('id', ''),
            x=data.get('x', 0.0),
            y=data.get('y', 0.0),
            radius=data.get('radius', 20.0),
            distance_to_piglet=data.get('distance_to_piglet', 0.0),
            is_piglet=data.get('is_piglet', False)
        )
    
    def __repr__(self) -> str:
        return f"Ball(id={self.id}, x={self.x:.1f}, y={self.y:.1f}, radius={self.radius:.1f}, is_piglet={self.is_piglet})"
